packages/labs/analyzer/src/lib/model.ts

```diff
@@ -97,25 +112,113 @@ export class Module {
    */
   readonly jsPath: PackagePath;
   /**
-   * A list of all Declaration models in this module.
+   * A map of names to models or model factories for all Declarations in this module.
    */
-  readonly declarations: Array<Declaration>;
+  private readonly _declarationMap: DeclarationMap;
   /**
-   * A set of all dependencies of this module
+   * Private storage for all declarations within this module, memoized
+   * in `get declarations()` getter.
+   */
+  private _declarations: Declaration[] | undefined = undefined;
+  /**
+   * A set of all dependencies of this module, as module absolute paths.
    */
   readonly dependencies: Set<AbsolutePath>;
   /**
    * The package.json contents for the package containing this module.
    */
   readonly packageJson: PackageJson;
+  /**
+   * A map of exported names to local declaration names or References, in
+   * the case of re-exported symbols.
+   */
+  private readonly _exportMap: ExportMap;
+  /**
+   * A list of module paths for all wildcard re-exports
+   */
+  private _finalizeExports: (() => void) | undefined;
 
   constructor(init: ModuleInit) {
     this.sourceFile = init.sourceFile;
     this.sourcePath = init.sourcePath;
     this.jsPath = init.jsPath;
     this.packageJson = init.packageJson;
-    this.declarations = init.declarations;
+    this._declarationMap = init.declarationMap;
     this.dependencies = init.dependencies;
+    this._exportMap = init.exportMap;
+    this._finalizeExports = init.finalizeExports;
+  }
+
+  /**
+   * Ensures the list of exports includes the names of all reexports
+   * from other modules.
+   */
+  private _ensureExportsFinalized() {
+    if (this._finalizeExports !== undefined) {
+      this._finalizeExports();
+      this._finalizeExports = undefined;
+    }
+  }
+
+  /**
+   * Returns names of all exported declarations.
+   */
+  get exportNames() {
+    this._ensureExportsFinalized();
+    return Array.from(this._exportMap.keys());
+  }
+
+  /**
+   * Given an exported symbol name, returns a Declaration if it was
+   * defined in this module, or a Reference if it was imported from
+   * another module.
+   */
+  getExport(name: string): Declaration | Reference {
+    this._ensureExportsFinalized();
+    const exp = this._exportMap.get(name);
+    if (exp instanceof Reference) {
+      return exp;
+    } else {
+      return this.getDeclaration(name);
+    }
+  }
+
+  /**
+   * Given an exported symbol name, returns the concrete Declaration
+   * for that symbol, following it through any re-exports.
+   */
+  getResolvedExport(name: string): Declaration {
+    let exp = this.getExport(name);
+    while (exp instanceof Reference) {
+      exp = exp.dereference();
+    }
+    return exp as Declaration;
+  }
+
+  /**
+   * Returns a top-level Declaration model for the given name.
+   */
+  getDeclaration(name: string) {
```

Comment:
Not needed. A module's declarations are only the symbols actually defined in the module (names local to the module). Exports (the actual interface for the module) can point to local declarations or references to exports of other modules (hence the need to resolve exports but not declarations).

We _might_ want to consider changing this down the road if/when we add support for class expressions and function expressions in variable initializers. `export class Foo {}` is a class declaration, but `export const Foo = class {}` is a variable declaration with a class expression initializer, but to the importer the difference is (mostly) unobservable, but only the former would get good class docs in the manifest. So we may want to treat the latter as a `ClassDeclaration` rather than a `VariableDeclaration`. Once we do that, then `import {Foo} from '...'; export const Bar = Foo;` is effectively a re-export in declaration form, so we might want to have `Declaration`s that are `Reference`s as well, or just elide the declaration for that altogether and just make it an export to a reference. TBD, @justinfagnani for thoughts.